File: app/src/main/java/com/worldventures/dreamtrips/social/ui/friends/service/command/AddFriendCommand.java
package com.worldventures.dreamtrips.social.ui.friends.service.command;

import com.worldventures.core.janet.CommandWithError;
import com.worldventures.core.janet.dagger.InjectableAction;
import com.worldventures.core.model.User;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.friends.SendFriendRequestHttpAction;

import javax.inject.Inject;

import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class AddFriendCommand extends CommandWithError<User> implements InjectableAction {

   private User user;
   private String circleId;

   @Inject Janet janet;

   public AddFriendCommand(User user, String circleId) {
      this.user = user;
      this.circleId = circleId;
   }

   @Override
   protected void run(CommandCallback<User> callback) throws Throwable {
      janet.createPipe(SendFriendRequestHttpAction.class)
            .createObservableResult(new SendFriendRequestHttpAction(user.getId(), circleId))
            .subscribe(action -> {
               callback.onSuccess(user);
            }, callback::onFail);
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.error_failed_to_send_friend_request;
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/service/users/search/command/AddFriendCommand.kt
package com.worldventures.dreamtrips.social.service.users.search.command

import com.worldventures.core.janet.CommandWithError
import com.worldventures.core.model.User
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.friends.SendFriendRequestHttpAction
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

@CommandAction
class AddFriendCommand(val user: User, val circleId: String) : CommandWithError<User>(), InjectableAction {

   @Inject lateinit var janet: Janet

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<User>) {
      janet.createPipe(SendFriendRequestHttpAction::class.java)
            .createObservableResult(SendFriendRequestHttpAction(user.id, circleId))
            .subscribe({ callback.onSuccess(user) }, callback::onFail)
   }

   override fun getFallbackErrorMessage() = R.string.error_failed_to_send_friend_request
}
